src/keys/mod.rs

```diff
@@ -94,6 +99,108 @@ impl<Ctx: ScriptContext> DescriptorKey<Ctx> {
         }
     }
 
+    /// Generate an `ExtendedPrivKey` based `DescriptorKey` for a given network with the default
+    /// generate options and random entropy. The key will be a master xkey with no origin or
+    /// extended derivation path.
+    pub fn generate_xprv(network: Network) -> Result<Self, KeyError> {
+        let mut xprv = <ExtendedPrivKey as GeneratableDefaultOptions<Ctx>>::generate_default()?;
+        //dbg!(&xprv);
+        xprv.key.network = network;
+        xprv.into_descriptor_key(None, DerivationPath::default())
+            .map(|k| match network {
+                Network::Bitcoin => k.override_valid_networks(mainnet_network()),
+                _ => k.override_valid_networks(test_networks()),
+            })
+    }
+
+    #[cfg(feature = "keys-bip39")]
+    /// Restore an `ExtendedPrivKey` based `DescriptorKey` for a given network from `Mnemonic`
+    /// seed words and optional password (BIP-39). The key will be a master xkey with no origin or
+    /// extended derivation path.
+    pub fn restore_xprv(
+        network: Network,
+        mnemonic: Mnemonic,
+        password: Option<String>,
+    ) -> Result<Self, KeyError> {
+        let xkey: ExtendedKey = (mnemonic, password).into_extended_key()?;
+        xkey.into_xprv(network)
+            .unwrap()
+            .into_descriptor_key(None, DerivationPath::master())
+    }
+
+    /// Derive a new `DescriptorKey` from the origin key at a given `DerivationPath` keeping the
+    /// extended path part the same.
+    pub fn derive(&self, path: DerivationPath) -> Result<Self, KeyError> {
+        let secp = Secp256k1::new();
+        match self {
+            DescriptorKey::Public(DescriptorPublicKey::XPub(xpub), _, _) => {
```

Comment:
I think you should not ignore the second item in the tuple (which is the set of valid networks). When you derive a key, the valid networks don't change, so you should use them in the returned key.

Also maybe this should be implemented on the `ExtendedKey` type instead? To avoid having to deal with unsupported key types, because `DescriptorKey` can also contain non-derivable keys.